docs/bsp.md

```diff
@@ -742,6 +760,155 @@ trait ResourcesItem {
 }
 ```
 
+### Task Notifications
+
+The server may inform the client on the state of running tasks, such as compilation or tests.
```

Comment:
Do you think it's strictly necessary to send the reports? I'm thinking the simplest implementation will just reply with a failed `CompileResult` etc and let the user read messages for details

Reply:
I think we should make it compulsory to send the reports for compilation and test, yeah. They can be seen as a hitting a "milestone" in compilation and clients can benefit from having it, especially when several modules are compiled for a given compile action.
  
I don't think there's a use case for making them optional, and trivially all build tools can send compilation reports (that's what all of them do whenever they print to console errors and warnings).

Reply:
I'm thinking for example of a maximally simple bsp wrapper for some command line tool. It would parse exit codes of command line compile invocations and otherwise just forward messages. Or just a really lazy build tool implementor that wants a quick IDE integration while hacking on the tool implementation. I don't think there's a correctness reason to strictly require these notifications.

Reply:
I think we should focus 

> I don't think there's a correctness reason to strictly require these notifications.
  
I agree, I haven't done any correctness reason. What I'm claiming is that:
  
1. This is not a point of friction for build tools. Giving this information is simple and should be forced in the protocol.
2. It makes the life of the IDE clients useful. Clients can (and probably need to) use the notification as a milestone after which they do a concrete action, show a popup or add to a view, etc.
  
I cannot imagine how a maximally simple bsp wrapper would be even able to support the rest of our non-trivial build tool requirements in the protocol, so I think this is not a use case we should support.

Reply:
> A plugin that runs some tasks to extract build structure, can run particular tasks and report success/failure.
  
It is a point of friction because you're not defining it in the build. In this scenario, you cannot even comply with the BSP connection protocol or other requirements in the protocol.
  
I'm inclined to not support this use case. A BSP server implementor should at least have access to the full universe of a build tool, it just cannot be implemented as a plugin.

Reply:
> In this scenario, you cannot even comply with the BSP connection protocol or other requirements in the protocol.

Why not?

I think the discussion is sort of moot anyway. If a feature is not technically necessary for other features to work, implementors may or may not bother with it, whatever the spec says. I just think we better acknowledge this in the spec directly.

Reply:
> Before this PR, sending CompileReport and TestReport was mandatory. I believe we should stick to the same semantics. I would prefer to discuss whether we should make this change or not in another PR, so that we don't block merging.

Sure. I made it optional to be in-line with task progress notifications in general, but agree to discuss in separate ticket.